vue-language-tools/vue-component-meta/README.md

```diff
@@ -1,10 +1,195 @@
 # vue-component-meta
 
-## Usage
+`vue-component-meta` allows you to extract the meta-data like props, slots, events, etc from your components via static code analysis. You can even generate description for your props from your source code. This helps document your components via automation. Please refer to the [reference](#reference) section for references.
 
-See https://github.com/johnsoncodehk/volar/blob/master/vue-language-tools/vue-component-meta/tests/index.spec.ts.
+## Guide 📗
 
-## Sponsors
+First of all, you need to create a component meta checker using `createComponentMetaChecker`:
+
+```ts
+import * as url from 'url'
+import path from 'path'
+
+import type { MetaCheckerOptions } from 'vue-component-meta'
+import { createComponentMetaChecker } from 'vue-component-meta'
+
+const __dirname = url.fileURLToPath(new URL('.', import.meta.url))
+
+const checkerOptions: MetaCheckerOptions = {
+  forceUseTs: true,
+  schema: { ignore: ['MyIgnoredNestedProps'] },
+  printer: { newLine: 1 },
+}
+
+const tsconfigChecker = createComponentMetaChecker(
+  // Write your tsconfig path
+  path.join(__dirname, 'path-to-tsconfig')
+  checkerOptions,
+)
+```
+
+Now, you can extract the component meta using `getComponentMeta` method of checker:
+
+```ts
+import * as url from 'url'
+import path from 'path'
+
+const __dirname = url.fileURLToPath(new URL('.', import.meta.url))
+
+const componentPath = path.join(__dirname, 'path-to-component');
+const meta = checker.getComponentMeta(componentPath);
+```
+
+This meta contains really useful stuff like component props, slots, events and more. You can refer to its [type definition](https://github.com/johnsoncodehk/volar/blob/master/vue-language-tools/vue-component-meta/src/types.ts) for more details.
+
+### Extracting prop meta
+
+`vue-component-meta` will automatically extract the prop details like its name, default value, is required or not, etc. Additionally, you can even write prop description in source code via [JSDoc](https://jsdoc.app/) comment for that prop.
+
+```ts
+/**
+ * Hide/Show alert based on v-model value
+ */
+modelValue: {
+  type: Boolean,
+  default: null,
+},
+```
+
+When you extract the component meta and extract the `description` property of that prop it will be "Hide/Show alert based on v-model value" 😍
+
+> **Warning**
+>
+> Do note that `meta.props` will be array of props so you can't access it via `meta.props.<prop-name>`. Moreover, `meta.props` will also contain some global prop which you can identify via `prop.global` property.
+
+You can use it to document your component as you build your project without writing additional documentation.
+
+### Defining type in JavaScript projects
+
+You can use JSDoc's `@type` to define the prop type in JavaScript projects.
+
+```ts
+props: {
+  /*
+   * Array of texts to render as a message
+   * @type {import('vue').PropOptions<string[]>}
+  */
```

Comment:
Sorry to be a pain but proper JSDoc comment would be formatted like:

```suggestion
  /**
   * Array of texts to render as a message
   * @type {import('vue').PropOptions<string[]>}
   */
```

Reply:
Also I wonder if generated types are correct in this case....

Reply:
I haven't tried it as I am on TS projects always.

Do you want me to check on JS, Maybe I can do that tomorrow we will be leaving office soon.

Sorry for that JSDoc style.

Reply:
I'm sorry but this doesn't seem to work. I've tried it on tests in this repo and for prop defined this way:

```
		/**
		 * Default function Array
		 * @type {import('vue').Prop<[1, 2, 3]>}
		 */
		arrayDefault: {
			type: Array,
			default: () => [1, 2, 3]
		},
```

gotten results like:

```
    {
      "name": "arrayDefault",
      "global": false,
      "description": "Default function Array",
      "tags": [
        {
          "name": "type",
          "text": "{import('vue').Prop<number[]>}"
        }
      ],
      "required": false,
      "type": "unknown[] | undefined",
      "schema": {
        "kind": "enum",
        "type": "unknown[] | undefined",
        "schema": [
          "undefined",
          {
            "kind": "array",
            "type": "unknown[]",
            "schema": [
              "unknown"
            ]
          }
        ]
      },
      "default": "[1, 2, 3]"
    },
```

So the resulting type is `unknown[]` which means the JSDoc annotation is not taken into account.

(I've used `Prop` type here as `PropOptions` doesn't seem to be exported in Vue 3.x anymore.)

So it's best to just remove this case from the readme. Sorry for the noise.